### DISCUSSION TITLE
Trivy scan on selected files

### DISCUSSION THREAD

**Discussion by pohanhuangtw** [aquasecurity/trivy#8173] [Q&A] [answered]
### Issues
I want to scan only the go.mod file in the sbomTrivy folder using Trivy. However, instead of scanning just that file, Trivy scans the entire directory.

### Context
The directory structure is as follows:
```bash
.
├── go.mod
├── go.sum
├── main.go
├── sbomTrivy
│   ├── go.mod
│   ├── go.sum
│   └── yy.go
```

I want to scan only sbomTrivy/go.mod.

- When I run trivy fs sbomTrivy, it correctly scans only the sbomTrivy directory.

- But when I run trivy fs --file-patterns "gomod:./sbomTrivy/go.mod" ., it scans the entire directory.

### Question
Why does using trivy fs --file-patterns "gomod:./sbomTrivy/go.mod" . cause Trivy to scan the whole directory instead of just the go.mod file in sbomTrivy?

### Target

Filesystem

### Scanner

None

### Output Format

None

### Mode

None

### Operating System

mac

### Version

```bash
0.58
```

**Comment by nikpivkin** [comment]
Hi @pohanhuangtw !

The `file-patterns` flag is used to look for files according to a pattern for a specific analyzer (see https://trivy.dev/latest/docs/configuration/skipping/#file-patterns). If you want to scan only `sbomTrivy/go.mod` then use it as a target.

**Reply by nikpivkin** [reply]
@knqyf263 [This](https://github.com/aquasecurity/trivy/blob/49c54b49c6563590dd82007d52e425a7a4e07ac0/docs/docs/configuration/skipping.md?plain=1#L104) link in the `file patterns` section points to custom checks, which seems wrong. I have not found a list of patterns in the documentation to update the link.

**Reply by pohanhuangtw** [reply]
Hi @nikpivkin ,
Thanks for the quick response.

If I have multiple go.mod files under the current directory and I want all of them included in a single SBOM—for example, from the directories sbomTrivy and sbomTrivy2—do I need to run Trivy twice and then merge the resulting SBOMs, or is there a way to include all go.mod files in a single scan (not the go.mod in the . direcotory)?

like the following
```bash
.
├── go.mod
├── go.sum
├── main.go
├── sbomTrivy
│   ├── go.mod
│   ├── go.sum
│   └── yy.go
├── sbomTrivy2
│   ├── go.mod
│   ├── go.sum
│   └── yy.go

```

**Reply by nikpivkin** [reply]
You can use the `--skip-files` flag:

```bash
❯ tree
.
├── foo
│   ├── go.mod
│   ├── go.sum
│   └── main.go
├── go.mod
├── go.sum
└── main.go

❯ trivy fs . -q -f json | jq '.Results[].Target'
"foo/go.mod"
"go.mod"

❯ trivy fs . -q -f json --skip-files go.mod | jq '.Results[].Target'
"foo/go.mod"
```

**Reply by pohanhuangtw** [reply]
From my understanding, the following command:
```bash
 fs --file-patterns "gomod:./sbomTrivy/go.mod" .
```
should search and scan `./sbomTrivy/go.mod` only, Why, then, does it also include the `go.mod` file from the current directory in the scan results?

**Reply by nikpivkin** [reply]
From the documentation:
> <b>In addition to the default file patterns</b>, the --file-patterns option takes regexp patterns to look for your files

**Reply by pohanhuangtw** [reply]
If the directory structure becomes more complex—for example, it has subdirectories named a, b, c, d, and so on—and you only want to scan certain directories (like b and c, or sometimes just a and b)

Ｈow can you specify exactly which files or directories should be scanned? Is there a straightforward way to list them explicitly, or does Trivy have any feature that allows selective scanning of specific paths?

**Reply by nikpivkin** [reply]
You can specify only one directory to scan at a time

/cc @knqyf263

**Reply by pohanhuangtw** [reply]
Based on my requirements, I might need to scan multiple different files. Would you recommend scanning each one individually and then merging the results (including generating the SBOM)?

Thanks for the answering.

**Reply by knqyf263** [reply]
This may be relevant.
https://github.com/aquasecurity/trivy/pull/8082

**Reply by pohanhuangtw** [reply]
Hi @knqyf263,
Thank you for your help. I think the PR is pretty close to what I was expecting, but the main goal is to focus on the path itself, not just the directory.

**Reply by pohanhuangtw** [reply]
After look into the [PR](https://github.com/aquasecurity/trivy/blob/5b93a9a9dddd6575ca52fcbf47b33c3de76c5ac5/pkg/fanal/walker/fs.go#L69)

opt.OnlyDirs can indeed apply to directories; however, this line appears to be used for files as well. 
The OnlyPath function is used to determine if the given path (whether file or directory) should be included based on the patterns in opt.OnlyDirs.

How about suggesting rename `opt.OnlyDirs` to `opt.OnlyPaths` to indicate it applies to paths (files or directories) rather than directories exclusively.

Will Trivy consider this approach in next release?

**Reply by knqyf263** [reply, marked as answer]
Actually, introducing `--only-dirs` in addition to `--skip-dirs` complicates usage. We need careful discussion about this enhancement. We cannot promise it will be included in the next version.

**Comment by silverwind** [comment]
This still seems a bit unwieldly to use because `trivy fs` only allows a single target. How about allowing multiple targets to be specified? e.g.:

```
trivy fs [flags] PATH...
```

For example, I have 3 potential target files but only want to scan 2 of them. I think the best way to specify this would be by just specifying them via `trivy fs uv.lock package-lock.json`.
